src/main.jl

```diff
@@ -327,11 +329,12 @@ function resolve(goals::Vector{<:Term}, clauses::ClauseTable{T}; options...)  wh
     search = get(options, :search, :bfs)::Symbol
     vcount = get(options, :vcount, Ref(UInt(0)))::Ref{UInt}
     # Construct top level goal and put it on the queue
-    queue = [GoalTree(Const(false), nothing, Vector{Term}(goals), 1, env, Subst())]
-    subst = []
+    queue = MutableLinkedList{GoalTree}()
```

Comment:
Is there a reason why you switched to using MutableLinkedLists? I believe Julia arrays have (amortized) constant time for pushes and pops at both the start and end of the array (see [this thread](https://discourse.julialang.org/t/performance-implications-of-array-mutations-push-splice-pop-etc/5482)), so I would only want to switch to a MutableLinkedList if there are clear performance gains on the existing benchmarks.

Reply:
I think if we're going to change data structures (which I'd do as a separate PR), it seems better to use a `Deque`, which appears to have faster performance for `pushfirst!` and `popfirst!`: https://juliacollections.github.io/DataStructures.jl/latest/deque/

However, I think this doesn't really matter for DFS, since we can rewrite everything to use only `push!` and `pop!` from the end of a `Vector`.